### PR TITLE
[cxx-interop] Synthesize conformances to `UnsafeCxxInputIterator`

### PR DESCRIPTION
**Pull request by egorzhdan**
This teaches ClangImporter to synthesize conformances of C++ iterator types to `UnsafeCxxInputIterator` protocol from the `Cxx` module.

We consider a C++ type to be an iterator if it defines a subtype (usually a typedef or a using decl) called `iterator_category` that inherits from `std::input_iterator_tag`.

rdar://96235368

### CODE REVIEW

File: lib/ClangImporter/ClangDerivedConformances.cpp [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+//
+//===----------------------------------------------------------------------===//
+
+#include "ClangDerivedConformances.h"
+#include "swift/AST/NameLookup.h"
+
+using namespace swift;
+
+void swift::conformToCxxIteratorIfNeeded(
+    ClangImporter::Implementation &impl, StructDecl *decl,
```
Comment (zoecarver) [inline lib/ClangImporter/ClangDerivedConformances.cpp:19]: Would we ever want this to work for foreign reference types? 
```suggestion
    ClangImporter::Implementation &impl, NominalTypeDecl *decl,
```

File: lib/ClangImporter/ImportDecl.cpp [hunk 2]
```diff
@@ -2470,7 +2471,18 @@ namespace {
...
         return nullptr;
 
-      return VisitRecordDecl(decl);
+      auto result = VisitRecordDecl(decl);
+
+      // If this module is declared as a C++ module, try to synthesize
+      // conformances to Swift protocols from the Cxx module.
+      auto clangModule = decl->getOwningModule();
+      if (clangModule && requiresCPlusPlus(clangModule)) {
+        if (auto structDecl = dyn_cast_or_null<StructDecl>(result)) {
```
Comment (zoecarver) [inline lib/ClangImporter/ImportDecl.cpp:2480]: Same as above.

**Review by zoecarver — APPROVED** [review]
This is so awesome!! Thanks Egor.

File: lib/ClangImporter/ClangDerivedConformances.cpp (hunk 1 above)
Comment (egorzhdan) [reply lib/ClangImporter/ClangDerivedConformances.cpp]: Yeap, you're right, thanks!

**Comment by egorzhdan** [comment]
@swift-ci please smoke test

File: lib/ClangImporter/ClangDerivedConformances.cpp [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+    return;
+  auto successor = dyn_cast<FuncDecl>(successors.front());
+  if (!successor || successor->isMutating())
+    return;
+  auto successorTy = successor->getResultInterfaceType();
+  if (!successorTy || successorTy->getAnyNominal() != decl)
+    return;
+
+  // Check if present: `func ==`
+  // FIXME: this only detects `operator==` declared as a member.
```
Comment (egorzhdan) [inline lib/ClangImporter/ClangDerivedConformances.cpp:96]: I will address this in a separate patch.

Comment (Huddie) [reply lib/ClangImporter/ClangDerivedConformances.cpp:96]: What does this mean?

**Comment by egorzhdan** [comment]
@swift-ci please smoke test macOS

**Comment by egorzhdan** [comment]
@swift-ci please smoke test

**Pull request merged**